### DISCUSSION TITLE
Tag, ITagRepository & Tag property type improvements

### DISCUSSION THREAD

**Discussion by drummerboy46** [umbraco/Umbraco-CMS#15840] [Features and ideas]
Taxonomy tagging is a really important part of Umbraco and for content management in general, but I feel Tag implementation in Umbraco vs using pre-defined content nodes and a MNTP is lacking. The problem for content editors, is you cannot expect them to remember taxonomy naming and Umbraco tag functionality requires them to essentially "make" tags up (even the same tags) as they create new content, This is a problem for cross content relationships if choosing to use the Tag datatype.

Here's some thoughts...

1. It's entirely possible I've missed something, but is ITagRepository the only API to manipulate Tags? It seems to be missing things? Would it be considered "normal" to have to implement separate Indexing code for the Assign() function? If you're programmatically assigning tags to a property would it not make sense to add the tags to the standard internal/external index? Or offer a method to do so?
2. In ITagRepository can we have some kind of ability to Create, Delete and Update Tags with group (or default) assignments? Create by default would have no assignment to any content properties. I cannot find an API to do this and have to revert to modifying the database tables! It seems a little crazy! 
3. Can we have a pre-value switch in the Datatype for Tags, that will enable a GetAllUnassignedTags by group as well as the tags used in the content property typeahead?
4. Can we have a pre-value switch in the Datatype for Tags that makes the typeahead read-only, i.e., it wont allow you to make up new tags (can be part of item 3) only select existing.
5. Can we have another Tag datatype, that doesn't require the content editor to "remember" tag names in a typeahead, providing a simple multi-select "TagList" that you can choose tags from a group?
6. A backoffice Tag Manager!

Again apologies if i've missed APIs for this, but I couldn't find anything in the source or docs.

I've been slowly improving this https://github.com/usome/UmbracoTagManager usome/huws TagManager mostly to the underlying management code, but a lack methods in ITagRepository makes database access the only way to achieve the Manager.

I've also created a TagList DataType that reads all tags assigned to a group and lets you pick them. The tags are indexed in internal/external indexes as comma delimited strings and stored by default in the property as Json.

![Screenshot 2024-03-05 124359](https://github.com/umbraco/Umbraco-CMS/assets/6703035/2ce21c33-c912-452e-b28e-2b68a619d4ea)

![Screenshot 2024-03-05 124449](https://github.com/umbraco/Umbraco-CMS/assets/6703035/8d4074b1-ccc7-46fa-8c9e-f43c3ef24b4d)

**Comment by melvinvanrookhuizen** [comment]
@drummerboy46 thanks for your input! I am completely clueless on the technical side, my team will get back to you shortly.

**Reply by nul800sebastiaan** [reply]
Not sure what your team needs to do @melvinvanrookhuizen - this was feedback for Umbraco HQ! 😉

**Comment by nul800sebastiaan** [comment]
@drummerboy46 just a reply (not an "official HQ reply") from skimming your suggestion: I believe the Contentment package probably covers your needs at the moment, I would specifically check out the Data List editor -https://github.com/leekelleher/umbraco-contentment / https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md

**Reply by drummerboy46** [reply]
@nul800sebastiaan yea Contentment is cool and I use it in my projects, but i'm mostly suggesting significant improvements to Tag APIs in Umbraco as well as property editor improvements around Tag use.

**Reply by nul800sebastiaan** [reply]
Understood! Please note that tag management is not currently on our roadmap to look at in the near future, we recommend using packages (and I'm sure with some work you can even make Contentment look at the Umbraco Tags as well). I'll put this on the `ideas` list for when we get to  work on tag related topics again.

**Comment by melvinvanrookhuizen** [comment]
Oh, haha. I though it was for our Tag Manager package..




Korte afspraak inplannen? Gebruik deze link naar mijn agenda!<https://calendly.com/zaaks/>



Melvin van Rookhuizen

https://nl.linkedin.com/in/melvinvanrookhuizen

06-15909546

 [1479189872460_handtekening]

Maliebaan 24-26

3581CB  Utrecht

www.zaaks.nl<http://www.zaaks.nl>

030-2769146

________________________________
Van: Sebastiaan Janssen ***@***.***>
Verzonden: dinsdag 5 maart 2024 14:44
Aan: umbraco/Umbraco-CMS ***@***.***>
CC: Melvin van Rookhuizen ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [umbraco/Umbraco-CMS] Tag, ITagRepository & Tag property type improvements (Discussion #15840)


Not sure what your team needs to do @melvinvanrookhuizen<https://github.com/melvinvanrookhuizen> - this was feedback for Umbraco HQ! 😉

**Reply by drummerboy46** [reply]
@melvinvanrookhuizen  I've commented on this specifically regarding TagManager in this issue https://github.com/usome/UmbracoTagManager/issues/11
